File: Renderer/src/main/java/darwin/resourcehandling/shader/ShaderLoader.java
/*
 * Copyright (C) 2012 some
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package darwin.resourcehandling.shader;

import java.io.*;
import java.nio.file.Path;
import java.nio.file.Paths;

import darwin.renderer.GraphicContext;
import darwin.renderer.opengl.*;
import darwin.renderer.shader.Shader.ShaderFactory;
import darwin.renderer.shader.*;
import darwin.resourcehandling.dependencies.annotation.InjectBundle;
import darwin.resourcehandling.factory.ResourceFromBundle;
import darwin.resourcehandling.handle.*;
import darwin.resourcehandling.shader.ShaderFile.Builder;
import darwin.util.logging.InjectLogger;

import com.google.common.base.Optional;
import javax.inject.Inject;
import javax.media.opengl.*;
import org.slf4j.Logger;
import org.slf4j.helpers.NOPLogger;

import static darwin.renderer.opengl.ShaderType.*;

/**
 *
 * @author some
 */
public class ShaderLoader implements ResourceFromBundle<Shader> {

    @InjectBundle(files = {"Empty.frag", "Empty.vert"}, prefix = SHADER_PATH_PREFIX)
    private ResourceBundle empty;
    @InjectLogger
    private Logger logger = NOPLogger.NOP_LOGGER;
    public static final String INCLUDE_PREFIX = "#pragma include";
    public static final String SHADER_PATH_PREFIX = "resources/shaders/";
    public static final Path SHADER_PATH = Paths.get(SHADER_PATH_PREFIX);
    private final ShaderObjektFactory soFactory;
    private final GLClientConstants constants;
    private final FileHandlerFactory fileFactory;
    private final ShaderFactory shaderFactory;
    private final GraphicContext gcontext;

    @Inject
    public ShaderLoader(ShaderObjektFactory soFactory, GLClientConstants constants,
                        FileHandlerFactory fileFactory, ShaderFactory shaderFactory, GraphicContext gcontext) {
        this.soFactory = soFactory;
        this.constants = constants;
        this.fileFactory = fileFactory;
        this.shaderFactory = shaderFactory;
        this.gcontext = gcontext;
    }

    @Override
    public Shader create(ResourceBundle bundle) throws IOException {
        Builder builder = new Builder();
        switch (bundle.getCount()) {
            default:
            case 3:
                builder.withGeometrie(getData(bundle.get(2).getStream()));
            case 2:
                builder.withVertex(getData(bundle.get(1).getStream()));
            case 1:
                builder.withFragment(getData(bundle.get(0).getStream()));
        }
        builder.withName(bundle.toString()).withMutations(bundle.getOptions());

        final ShaderFile file = builder.create();
        final Shader shader = shaderFactory.create(file);

        gcontext.invoke(false, new GLRunnable() {
            @Override
            public boolean run(GLAutoDrawable glad) {
                try {
                    ShaderProgramm compiledShader = compileShader(file);
                    ShaderProgramm old = shader.getProgramm();
                    shader.ini(compiledShader);
                    logger.info("Shader " + file.name + " was succesfully compiled!");
                    if (old != null) {
                        old.delete();
                    }
                } catch (Throwable ex) {
                    logger.warn(ex.getLocalizedMessage());
                }
                return true;
            }
        });

        return shader;
    }

    @Override
    public void update(final ResourceBundle bundle, ResourceHandle changed, final Shader shader) {
        try {
            String d = null;
            ShaderType t = null;
            switch (bundle.getCount()) {
                default:
                case 3:
                    if (bundle.get(2) == changed) {
                        d = getData(bundle.get(2).getStream());
                        t = ShaderType.Geometrie;
                        break;
                    }
                case 2:
                    if (bundle.get(1) == changed) {
                        d = getData(bundle.get(1).getStream());
                        t = ShaderType.Vertex;
                        break;
                    }
                case 1:
                    if (bundle.get(0) == changed) {
                        d = getData(bundle.get(0).getStream());
                        t = ShaderType.Fragment;
                        break;
                    }
            }

            final ShaderType type = t;
            final String data = d;
            gcontext.invoke(false, new GLRunnable() {
                @Override
                public boolean run(GLAutoDrawable glad) {
                    try {
                        GL2GL3 gl = glad.getGL().getGL2GL3();
                        ShaderObjekt so = createSObject(type, data, bundle.getOptions());
                        int po = shader.getProgramm().getPObject();

                        gl.glAttachShader(po, so.getShaderobjekt());
                        gl.glLinkProgram(po);
                        shader.ini(shader.getProgramm());

                        Optional<String> error = shader.getProgramm().verify();
                        if (error.isPresent()) {
                            logger.warn(error.get());
                            shader.ini(getFallBack().getProgramm());
                        } else {
                            logger.info("Shader " + bundle + " was succesfully updated!");
                        }
                    } catch (Throwable ex) {
                        logger.warn(ex.getLocalizedMessage());
                    }
                    return true;
                }
            });
        } catch (IOException ex) {
            logger.warn(ex.getLocalizedMessage());
        }
    }

    @Override
    public Shader getFallBack() {
        try {
            return create(empty);
        } catch (IOException ex) {
            throw new RuntimeException("Could not load fallback shader!");
        }
    }

    //TODO compile fehler vllcht auffangen, zumindestens im DEV mode?
    public ShaderProgramm compileShader(ShaderFile sfile) {
//        boolean error = false;
//        StringBuilder errorText = new StringBuilder("Shader ");
        String errorMessage = null;
        BuildException exception = null;
        ShaderObjekt fso, vso, gso;
        fso = gso = vso = null;

        try {
            if (sfile.fragment != null) {
                fso = createSObject(Fragment, sfile.fragment, sfile.mutations);
            }
        } catch (BuildException ex) {
            exception = ex;
            errorMessage = "in:Fragment of";
        }
        try {
            if (sfile.vertex != null) {
                vso = createSObject(Vertex, sfile.vertex, sfile.mutations);
            }
        } catch (BuildException ex) {
            exception = ex;
            errorMessage = "in: Vertex of";
        }

        try {
            if (sfile.geometrie != null) {
                gso = createSObject(Geometrie, sfile.geometrie, sfile.mutations);
            }
        } catch (BuildException ex) {
            exception = ex;
            errorMessage = "in: Geometrie of";
        }

        if (exception == null) {
            try {
                return new ShaderProgramBuilder().
                        with(sfile.getAttributs()).
                        with(fso).with(vso).with(gso).
                        link(gcontext);
            } catch (BuildException ex) {
                exception = ex;
                errorMessage = "while linking";
            }
        }
        throw new RuntimeException("Shader " + exception.getErrorType() + " ERROR " + errorMessage
                                   + " {" + sfile.name + "}\n" + exception.getMessage());
    }

    public ShaderObjekt createSObject(ShaderType target, String source,
                                      String... mut) throws BuildException {
        String[] sources = new String[2 + mut.length];
        sources[0] = constants.getGlslVersion();
        for (int i = 0; i < mut.length; ++i) {
            sources[i + 1] = "#define " + mut[i] + '\n';
        }

        sources[mut.length + 1] = source;
        return soFactory.create(target, sources);
    }

    public String getData(InputStream file) {
        if (file == null) {
            return null;
        }
        String out;
        StringBuilder sb = new StringBuilder();
        try {
            Reader fr = new InputStreamReader(file);
            BufferedReader br = new BufferedReader(fr);
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.startsWith(INCLUDE_PREFIX)) {
                    String path = line.substring(INCLUDE_PREFIX.length()).trim();
                    InputStream shader = fileFactory.create(SHADER_PATH_PREFIX + path).getStream();
                    if (shader != null) {
                        String src = getData(shader);
                        sb.append(src);
                        try {
                            shader.close();
                        } catch (IOException ex) {
                        }
                    }
                    continue;
                } else if (line.startsWith("#version")) {
                    continue;
                }
                sb.append(line).append('\n');
            }
        } catch (IOException ex) {
            logger.error("Fehler beim laden eines Shader Source Strings: "
                         + ex.getLocalizedMessage());
        }
        out = sb.toString();
        try {
            file.close();
        } catch (IOException ex) {
        }
        return out;
    }
}

File: Renderer/src/main/scala/darwin/resourcehandling/shader/ShaderLoader.scala
/*
 * Copyright (C) 2012 some
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package darwin.resourcehandling.shader

import java.io._
import java.nio.file._
import darwin.renderer.shader._
import darwin.resourcehandling.factory.ResourceFromBundle
import darwin.resourcehandling.handle._
import javax.media.opengl._
import darwin.resourcehandling.ResourceComponent
import darwin.renderer.{ GProfile, GraphicComponent }
import darwin.util.logging.LoggingComponent
import darwin.resourcehandling.shader.brdf.BrdfReader

/**
 *
 * @author some
 */
object ShaderLoaderComponent {
  val INCLUDE_PREFIX = "#pragma include"
  val BRDF_PREFIX = "#pragma brdf"
  val BRDF_PATH_PREFIX = "resources/brdfs/"
  val SHADER_PATH_PREFIX = "resources/shaders/"
  val SHADER_PATH = Paths.get(SHADER_PATH_PREFIX)
}

trait ShaderLoaderComponent extends LoggingComponent {
  this: ShaderComponent with ShaderObjektComponent with ShaderProgrammComponent with ResourceComponent with GraphicComponent with GProfile[GL2GL3] =>

  import ShaderLoaderComponent._

  implicit object ShaderLoader extends ResourceFromBundle[Shader] {

    def create(bundle: ResourceBundle): Shader = {
      val builder: ShaderFile.Builder = new ShaderFile.Builder

      val iter = Iterator.tabulate(bundle.getCount)(bundle get _)
      val facs = Iterator(builder.withFragment _, builder.withVertex _, builder.withGeometrie _)

      (iter zip facs) foreach (t => t._2(getData(t._1.getStream).get))

      builder.withName(bundle.toString).withMutations(bundle.getOptions: _*)
      val file: ShaderFile = builder.create
      val shader: Shader = createShader(file)

      context.asyncInvoke {
        glad =>
          val compiledShader = compileShader(file)

          for (cs <- compiledShader.left) {
            val old: ShaderProgramm = shader.getProgramm
            shader.ini(cs)
            logger.info("Shader " + file.name + " was succesfully compiled!")
            if (old != null) {
              old.delete
            }
          }
          for (ex <- compiledShader.right) {
            logger.warn("" + ex.getLocalizedMessage)
          }
      }

      return shader
    }

    def update(bundle: ResourceBundle, changed: ResourceHandle, shader: Shader) {
      val types = Seq(ShaderType.Fragment, ShaderType.Vertex, ShaderType.Geometrie)
      val which = (1 to 3).map(bundle.get).indexOf(changed)

      assert(which == -1, "Resource handle to update isn't part of the bundle")

      val `type`: ShaderType = types(which)
      for (data <- getData(bundle.get(which).getStream)) {
        context.asyncInvoke { glad =>
          log {
            val gl: GL2GL3 = glad.getGL.getGL2GL3
            val so: ShaderObjekt = createSObject(`type`, data, bundle.getOptions: _*)

            val prog = shader.getProgramm

            val sobjects = prog.shaderObjects
            for (old <- sobjects.find(_.shaderType == so.shaderType)) {
              prog.detach(old)
              old.delete
            }

            prog.attach(so)
            prog.link()

            val error = prog.verify
            if (error.isDefined) {
              logger.warn(error.get)
              shader.ini(getFallBack.getProgramm)
            } else {
              shader.ini(prog)
              logger.info("Shader " + bundle + " was succesfully updated!")
            }
          }
        }
      }
    }

    lazy val getFallBack: Shader = resource("Empty.frag", "Empty.vert")(SHADER_PATH_PREFIX)

    def compileShader(sfile: ShaderFile): Either[ShaderProgramm, BuildException] = {
      val sobjects = for (st <- ShaderType.values()) yield {
        val src = st.extract(sfile)

        try {
          Right(Option(src).map(createSObject(st, _, sfile.mutations: _*)))
        } catch {
          case ex: BuildException => {
            Left((st, ex))
          }
        }
      }

      val errors = sobjects.flatMap(_.left.toOption)
      if (!errors.isEmpty) {
        val es = errors.map(e => s"\t${e._2.getErrorType.name} error in ${e._1.name}: ${e._2.getMessage}")
        throw new RuntimeException(s"Error(s) in Shader {${sfile.name}}:\n" + es.mkString("\n"))
      }

      val sos = sobjects.flatMap(_.right.toOption).flatten
      import scala.collection.JavaConverters._
      linkShaderProgramm(sos, sfile.getAttributs.asScala: _*)
    }

    def createSObject(target: ShaderType, source: String, mut: String*): ShaderObjekt = {
      val sources = (glslVersion +: mut.map(m => s"#define $m\n") :+ source).toArray
      return createShaderObject(target, sources)
    }

    def getData(file: InputStream): Option[String] = {
      if (file == null) {
        return None
      }

      val lines = scala.io.Source.fromInputStream(file).getLines()
      var end: List[String] = Nil

      val prepared = (lines.filterNot(_ startsWith "#version").map {
        case l if l startsWith INCLUDE_PREFIX => {
          val path: String = l.substring(INCLUDE_PREFIX.length).trim
          getData(handle(SHADER_PATH_PREFIX + path).getStream)
        }
        case l if l startsWith BRDF_PREFIX => {
          val path: String = l.substring(BRDF_PREFIX.length).trim
          val stream: InputStream = handle(BRDF_PATH_PREFIX + path + ".brdf").getStream
          val brdf = BrdfReader.readBrdf(stream, path)
          stream.close()

          end :+= brdf.getCode

          brdf.getFunctionTemplate
        }
        case l => l
      } ++ end).mkString("\n")

      file.close()

      Some(prepared)
    }
  }



}